### ISSUE TITLE
Acc drop significantly during the last epoch of stage1

### ISSUE THREAD

**Issue by FANG-Xiaolin** [xingyizhou/pytorch-pose-hg-3d#16]
Hi Xingyi,
After training the 2D hourglass component for 50+ epochs, the accuracy is approximately 83%, but after the 60th epoch, the accuracy suddenly drop to 43%. 

Here's the log.
![image](https://user-images.githubusercontent.com/25632937/38814929-498b789e-41c5-11e8-82cc-4647cb631a09.png)

**Comment by xingyizhou** [comment]
Hi, 
  As far as I know, it should be caused by a pytorch internal bug in BN. You can comment `model.eval()` in testing to see if the validation acc gets better (but it still won't match the desired performance). The bug should not be reproducible. And re-train the network once more (better on another machine) should have different results. Or you can downgrade your pytorch version below 0.1.12, which is a version where I haven't met/ heard about this bug (but still not guaranteed). Please let me know if the above solutions help. Thanks!

**Comment by fxlfang** [comment]
I tried for another 3 times. The train acc is approximately 0.87 during the last epoch(the 60th epoch) but the validation acc changes every time and always lower than 0.50. The validation acc is around 0.80 in the 55th epoch so it seems that there is a sudden drop during the last epoch and I notice that the training loss gets slightly higher during the last epoch.

**Comment by xingyizhou** [comment]
Hi, 
  Thanks for reporting the problem. However I don't have other solutions yet and will keep looking into it. It might not be a bug of the code, since an isolated implementation of HourglassNet (I am not sure if the bug is from the network architecture) also has this problem (https://github.com/bearpaw/pytorch-pose/issues/33). People there suggest using learning rate 1e-4. You can have a try to see if the bug still exists.

**Comment by fxlfang** [comment]
Hi,
Thanks for your advice. Yes it works if using LR 1e-4. The val acc is 0.80+ in this way.

**Comment by xingyizhou** [comment]
Hi, 
  I have investigated this problem (on another project, while I can not reproduce the bug on this project). It seems it is caused by very large intermediate features (e.g. > 10000) before batch normalization. Then the train() mode is on, it will be normalized be itself so training is OK. But when eval() mode is on, a slight difference (of the intermediate feature) with the BN mean/std from training will results in large offsets for output. I don't know the causal of the problem but it looks mathematically reasonable. However, down-grading PyTorch version to 0.1.12 will eliminate the problem. Please notify me if you have any other observations on this bug. Thanks!

**Comment by fxlfang** [comment]
Hi, 
Yes I think it is reasonable. Sure I will notify you if I observe something new. Thanks for your reply!

**Comment by SsnL** [comment]
I see. 6 is still too small though. People usually use >128 with BN.

**Comment by xingyizhou** [comment]
Hi all, 
  As pointed by @[leoxiaobin](https://github.com/leoxiaobin), turn off cudnn of BN layer resolves the issue. It can be realized by set `torch.backends.cudnn.enabled = False` in `main.py`, which disables cudnn for all layers and slows down the training by about 1.5x time, or re-build pytorch from source by hacking cudnn in BN layers https://github.com/pytorch/pytorch/blob/e8536c08a16b533fe0a9d645dd4255513f9f4fdd/aten/src/ATen/native/Normalization.cpp#L46 .

**Comment by FANG-Xiaolin** [comment]
Get it. Thanks.

**Comment by xingyizhou** [comment]
Oh I still want this issue to be opened to wait for better solutions...

**Comment by FANG-Xiaolin** [comment]
Sure! My bad.

**Comment by wangg12** [comment]
@SsnL @xingyizhou Does this bug still exist with pytorch >= 1.0?

**Comment by ujsyehao** [comment]
@wangg12  I am doing experiments to observe if the bug exists in pytorch >= 1.0.

**Comment by qiangruoyu** [comment]
> @ wangg12 我正在做实验，以观察pytorch> = 1.0中是否存在该错误。

Can you meet this error when the version of pytorch >= 1.0

**Comment by yy9568** [comment]
@ujsyehao 你好，请问你的实验结果如何？

**Comment by sisrfeng** [comment]
> Hi all,
> As pointed by @[leoxiaobin](https://github.com/leoxiaobin), turn off cudnn of BN layer resolves the issue. It can be realized by set `torch.backends.cudnn.enabled = False` in `main.py`, which disables cudnn for all layers and slows down the training by about 1.5x time, or re-build pytorch from source by hacking cudnn in BN layers https://github.com/pytorch/pytorch/blob/e8536c08a16b533fe0a9d645dd4255513f9f4fdd/aten/src/ATen/native/Normalization.cpp#L46 .

torch.backends.cudnn.enabled = False` in `main.py` 
Should it be "torch.backends.cudnn.benchmark = False"?

If I have followed this step, I need not modify main.py, right? :
For other pytorch version, you can manually open torch/nn/functional.py and find the line with torch.batch_norm and replace the torch.backends.cudnn.enabled with False
